src/EditorFeatures/Core/ExternalAccess/IntelliCode/Api/IIntentProcessor.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.IntelliCode.Api
+{
+    internal interface IIntentProcessor
+    {
+        /// <summary>
+        /// For an input intent, computes the edits required to apply that intent and returns them.
+        /// </summary>
+        /// <param name="intentRequestContext">the intents with the context in which the intent was found.</param>
+        /// <returns>the edits that should be applied to the current snapshot.</returns>
+        Task<ImmutableArray<IntentResult>> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Defines the data needed to compute the code action edits from an intent.
+    /// </summary>
+    internal struct IntentRequestContext
+    {
+        /// <summary>
+        /// The intent name.  <see cref="WellKnownIntents"/> contains all intents roslyn knows how to handle.
+        /// </summary>
+        public string IntentName { get; }
+
+        /// <summary>
+        /// JSON formatted data specific to the intent that must be deserialized into the appropriate object.
+        /// </summary>
+        public string? IntentData { get; }
+
+        /// <summary>
+        /// The text snapshot and selection when <see cref="IIntentProcessor.ComputeEditsAsync"/>
+        /// was called to compute the text edits and against which the resulting text edits will be calculated.
+        /// </summary>
+        public SnapshotSpan CurrentSnapshotSpan { get; }
+
+        /// <summary>
+        /// The text edits that should be applied to the <see cref="CurrentSnapshotSpan"/> to calculate
+        /// a prior text snapshot before the intent happened.  The snapshot is used to calculate the actions.
+        /// </summary>
+        public ImmutableArray<TextChange> PriorTextEdits { get; }
+
+        /// <summary>
+        /// The caret position / selection in the snapshot calculated by applying
+        /// <see cref="PriorTextEdits"/> to the <see cref="CurrentSnapshotSpan"/>
+        /// </summary>
+        public TextSpan PriorSelection { get; }
```

Comment:
i think this api shape is overall fine for now.  i do think in teh future we should probably ask the platform/intellicode to provide a 'prior' ITextSnapshot.  

We'll probably need to doc that it will not be attached to an actual text buffer.